capacity4more/modules/c4m/features_og/c4m_features_og_events/c4m_features_og_events.module

```diff
@@ -219,38 +219,110 @@ function c4m_features_og_events_entity_info_alter(&$entity_info) {
  * 1 Event for anonymous user, 3 for logged in.
  *
  * @return string
- *  HTML of rendered upcoming events.
+ *  HTML of rendered upcoming events, FALSE if there's no events.
  */
 function c4m_features_og_events_upcoming_events() {
   global $user;
   $events = '';
+  $filter = '';
 
   // Show all events starting from midnight onwards.
   $tomorrow = new DateTime('midnight tomorrow');
   $date = $tomorrow->format('Y-m-d H:i:s');
 
-  // User's group when user is logged in, Public groups when not.
-  $groups = $user->uid ? og_get_groups_by_user() : c4m_og_get_groups_by_access_type('public');
+  // Get user's interests when user is logged in.
+  // Get content filter when user is logged in.
+  // Get user's group when user is logged in, Public groups when not.
   // 3 Events when user is logged in, 1 when not.
-  $range = $user->uid ? 3 : 1;
+  if ($user->uid) {
+    // Check if content filter is applied.
+    $user_wrapper = entity_metadata_wrapper('user', $user->uid);
+    $interests = $user_wrapper->c4m_related_topic->raw();
+    $filter = !empty($_GET['filter']) ? $_GET['filter'] : '';
+    $groups = array();
+    $range = 3;
+  }
+  else {
+    $groups = c4m_og_get_groups_by_access_type('public');
+    $range = 1;
+  }
+
+  // Filter: Case "interests": Public groups and restricted groups which user's mail address is matching the domain AND tagged with one or more of the users topics of interest.
+  // Filter: Case "groups": User's groups, If there's no groups, No events are shown.
+  // Filter: Default: Show all events from all groups.
+  switch ($filter) {
+    case 'interests':
+
+      $groups = og_get_groups_by_user();
+      $groups = !empty($groups['node']) ? $groups['node'] : array();
+
+      if (!empty($interests)) {
+        $query = new EntityFieldQuery();
+
+        $query
+          ->entityCondition('entity_type', 'node')
+          ->entityCondition('bundle', 'group')
+          ->propertyCondition('status', NODE_PUBLISHED)
+          ->fieldCondition('c4m_related_topic', 'target_id', $interests, 'IN');
+        if (!empty($groups)) {
+          $query->propertyCondition('nid', $groups, 'NOT IN');
+        }
+        $result = $query->execute();
+
+        $groups += !empty($result['node']) ? array_keys($result['node']) : array();
+      }
+      break;
+    case 'groups':
+      $groups = og_get_groups_by_user();
+      $groups = !empty($groups['node']) ? $groups['node'] : array('no_groups');
+      break;
+    default:
+      break;
```

Comment:
Yea i know, I thought maybe we can change the code to have this happen in the default of the switch.
Anyway. Looks good otherwise, i will assign to Peter